SP23.P03.Web/Controllers/TrainsController.cs

```diff
@@ -0,0 +1,153 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using SP23.P03.Web.Data;
+using SP23.P03.Web.Features.Authorization;
+using SP23.P03.Web.Features.Trains;
+
+namespace SP23.P03.Web.Controllers
+{
+    [Route("api/trains")]
+    [ApiController]
+    public class TrainsController : ControllerBase
+    {
+        private readonly DbSet<Train> trains;
+        private readonly DataContext dataContext;
+        //private readonly int id;
+
+        public TrainsController(DataContext dataContext)
+        {
+            this.dataContext = dataContext;
+            trains = dataContext.Set<Train>();
+        }
+
+        [HttpGet]
+        public IQueryable<TrainDto> GetAllTrains()
+        {
+            return GetTrainDtos(trains);
+        }//end GetAllTrains
+
+        [HttpGet("{id}")]
+        [Authorize]
+        public ActionResult<TrainDto> GetTrainById([FromRoute] int id)
+        {
+            var train = trains.FirstOrDefault(x => x.Id == id);
+
+            if (train == null)
+            {
+                return NotFound();
+            }
+
+            var trainDto = new TrainDto
+            {
+                Id = train.Id,
+                Name = train.Name,
+                Status = train.Status,
+                Capacity = train.Capacity
+            };
+
+            return Ok(trainDto);
+        }//end GetTrainById
+
+        [HttpPost]
+        [Authorize(Roles = RoleNames.Admin)]
+        public ActionResult<CreateTrainDto> CreateTrain(CreateTrainDto createTrainDto, int id)
+        {
+            var train = trains.FirstOrDefault(x => x.Id == id);
```

Comment:
```suggestion
        public ActionResult<TrainDto> CreateTrain(CreateTrainDto createTrainDto)
        {
```
1. This method returns a TrainDto rather than a CreateTrainDto, so that's what belongs in the return type. This doesn't cause any actual errors for some reason, but may as well fix it.
2. You don't need to get an id or find a train to create a train, so that code can just be removed.